### DISCUSSION TITLE
SQLModel does not execute type checks in pydantic when `table=True` set

### DISCUSSION THREAD

**Discussion by monchin** [fastapi/sqlmodel#1036] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel
from sqlmodel import SQLModel, Field

class TestPD(BaseModel):
    id: int
    name: str
    age: int

class TestSM(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    name: str
    age: int

def test_pd():
    test = TestPD(id=1, name="test", age="test")
    print(test)

def test_sm():
    test = TestSM(id=1, name="test", age="test")
    print(test)

# test_pd() # raise ValidationError here
test_sm() # doesn't raise
```

### Description

Hi, I just found that althogh `SQLModel` derives from `pydantic.BaseModel`, it doesn't really execute type checks in pydantic. Please read the code above. If we uncomment `test_pd()` and comment `test_sm()`, it would raise ValidationError saying `age` is not a valid integer; but if we execute `test_sm()`, everything goes well, output would be
```python
id=1 name='test' age='test'
```
although it should raise a ValidationError. And I found if I delete `table=True`, it behaves just like `BaseModel`, but in my condition `table=True` must be set.

I'd like to know why `pydantic` type checking doesn't work when I set `table=True`? I have found some simillar discussions for example [this](https://github.com/tiangolo/sqlmodel/discussions/811) but it remained unanswered. Is there a way to solve it?

### Operating System

Windows

### SQLModel Version

0.0.21

### Python Version

3.10.12

**Comment by monchin** [comment]
I read the source code, and found [if `table=True` is set](https://github.com/tiangolo/sqlmodel/blob/0.0.21/sqlmodel/_compat.py#L349), original pydantic validators will not executed, instead `sqlmodel_table_construct` will be excuted. That is the reason why pydantic doesn't work.

One reason I'm attrachted by `sqlmodel` is that it combines `pydantic` and orm, so I think it would be great if I can benefit all pydantic traits, not only for type checking, but also for example [`field_validator` to change values automatically](https://docs.pydantic.dev/latest/concepts/validators/#validation-of-default-values). Thus I tried to modify some code like this
```python
# source code
    def sqlmodel_init(*, self: "SQLModel", data: Dict[str, Any]) -> None:
        old_dict = self.__dict__.copy()
        if not is_table_model_class(self.__class__):
            self.__pydantic_validator__.validate_python(
                data,
                self_instance=self,
            )
        else:
            sqlmodel_table_construct(
                self_instance=self,
                values=data,
            )
# modified code
    def sqlmodel_init(*, self: "SQLModel", data: Dict[str, Any]) -> None:
        old_dict = self.__dict__.copy()
        raw_self = self.copy()
        pydantic_validated_model = self.__pydantic_validator__.validate_python(
            data,
            self_instance=self,
        )
        if is_table_model_class(self.__class__):
            sqlmodel_table_construct(
                self_instance=raw_self,
                values=pydantic_validated_model.model_dump(),
            )
```
In this way pydantic also works even though `table=True` is set, including type checking and all validators to validate or change values.
If you think it is a good idea and the code is OK, I'm glad to contribute!

**Comment by monchin** [comment]
Now I have fixed my code to pass tests
```python
    def sqlmodel_init(*, self: "SQLModel", data: Dict[str, Any]) -> None:
        # old_dict = self.__dict__.copy()
        # if not is_table_model_class(self.__class__):
        #     self.__pydantic_validator__.validate_python(
        #         data,
        #         self_instance=self,
        #     )
        # else:
        #     sqlmodel_table_construct(
        #         self_instance=self,
        #         values=data,
        #     )
        old_dict = self.__dict__.copy()
        if not is_table_model_class(self.__class__):
            self.__pydantic_validator__.validate_python(
                data,
                self_instance=self,
            )
        else:
            raw_self = deepcopy(self)
            pydantic_validated_model = self.__pydantic_validator__.validate_python(
                data,
                self_instance=raw_self,
            )
            pydantic_dict = pydantic_validated_model.model_dump()
            for k in pydantic_dict.keys():
                if k not in data.keys():
                    continue
                data[k] = pydantic_dict[k]
            sqlmodel_table_construct(
                self_instance=self,
                values=data,
            )
        object.__setattr__(
            self,
            "__dict__",
            {**old_dict, **self.__dict__},
        )
```
Not all tests passed, but tests do not pass also fails in original code, so I think my code works well. But for one test `tests/test_instance_no_args/test_allow_instantiation_without_arguments`, the original code passed but mine failed. In this test, I notice that original code deliberately initiate an instance of `SQLModel` without neccesary non-nullable parameters, and set the neccesary parameters after initiation. So I'd like to know if `table=True` disabling pydantic validators is a feature or a bug. I hope I can regard it as a bug because I hope to make full use of `pydantic` in `sqlmodel`, of course I can accept it if it is a feature.

**Comment by YuriiMotov** [comment]
There is related issue: https://github.com/fastapi/sqlmodel/issues/453

Let's close this discussion and track it there
